### PR TITLE
Add API config endpoints

### LINKED ISSUE #4822: Create new endpoint to manage API configuration
# Description

Currently, the only way to change any API setting is to edit the `api.yaml` file. This method forces users to login to the deployed server or use the `configure_api.sh` script. The goal of this issue is to extend API capabilities to handle its own configuration so everything could be done via REST requests.

Most of settings may be hot-changed without restarting the API daemon. They are quite unlikely to make the API service inaccessible, so we could call them 'soft settings':
- `behind_proxy_server`
- `rbac`.`mode`
- `logs`.`level`
- `cors`
- `cache`
- `use_only_authd`
- `experimental_features`

The rest of settings may need a restart of the service and cannot guarantee that the service is working again, so we could call them 'hard settings'. The only way to change this settings will be editing the `api.yaml` file. The reason is to ensure that if an administrator makes a mistake and breaks the environment, there is another chance to restore it:
- `host`
- `port`
- `https`
- `logs`.`path`
- `drop_privileges`

## Design

The following endpoints will be created to handle soft settings:
- Read access to configuration:
   - `GET /manager/api/config`
   - `GET /cluster/api/config`
- Selectively edit configuration:
   - `PUT /manager/api/config`
   - `PUT /cluster/api/config`
- Restore configuration to factory settings:
   - `DELETE /manager/api/config`
   - `DELETE /cluster/api/config`

Some use cases are:
- Get configuration
```json
# GET /manager/api/config
{
        "behind_proxy_server": false,
        "rbac": {
            "mode": "black"
        },
        "logs": {
            "level": "info"
        },
        "cors": true,
        "cache": {
            "enabled": false,
            "debug": false,
            "time": 0.750
        },
        "use_only_authd": false,
        "experimental_features": false
}
```
- Change setting
```json
# PUT /manager/api/config
{"cache": 
    {"enabled": false}
}
```
```json
# Response 200
{"message": "API configuration update successfully."
}
```
- Restore configuration:
```json
# DELETE /manager/api/config
{"message": "API configuration restored to factory settings."
}
```

Basically, the logic of these endpoints is to modify the `api.yaml` file in `/var/ossec/api/configuration`. The spec should be detailed enough to validate the body of the PUT endpoints that fits the configuration template perfectly.

### RBAC

The resource here is the already implemented `node:id:{node_id}`. See `manager.py` for a more detailed reference. However, new actions should be created for read, modifying and restoring the configuration.

## Tasks
- [x] Modify `spec.yaml` to include new endpoints
- [x] Add new functions in the framework to give support to the required logic and RBAC
- [x] Update RBAC catalogue
- [x] Review default roles
- [x] Write integration tests (basic, RBAC while-list and RBAC black-list)

### PR DESCRIPTION
**Pull request by Selutario**
This PR closes #4822 

# Description
This pull requests add a new endpoint which allows different actions related to the API configuration. Specifically, it is possible to query the current one (GET), modify it (PUT) or restore it to the default value (DELETE).

To avoid problems like modifying the IP and then not being able to access the API to return it to its original value, only some fields that do not interfere with the control and access of the API are available. The parameters that can be modified through the endpoint are the following:
- `behind_proxy_server` 
- `rbac`.`mode`
- `rbac`.`auth_token_exp_timeout`
- `logs`.`level`
- `cors`.`enabled`
- `cors`.`source_route`
- `cors`.`expose_headers`
- `cors`.`allow_headers`
- `cors`.`allow_credentials`
- `cache`.`enabled`
- `cache`.`time`
- `use_only_authd`
- `experimental_features`

The specification described in `spec.yaml` checks and prevents the introduction of parameters other than those mentioned. Also, the DELETE method only restores those parameters. This avoids accidentally changing the IP, port, etc. to default value.

# Tests performed
### Tests manager and RBAC
```
pytest test_manager_endpoints.tavern.yaml
=================================================== test session starts ===================================================
platform linux -- Python 3.7.5, pytest-5.4.1, py-1.8.1, pluggy-0.13.1
rootdir: /home/selu/Git/wazuh/api/test/integration, inifile: pytest.ini
plugins: tavern-1.0.0, cov-2.8.1
collected 35 items                                                                                                        

test_manager_endpoints.tavern.yaml ...................................                                              [100%]

==================================================== warnings summary =====================================================
```
```
pytest test_rbac_white_manager_endpoints.tavern.yaml
=================================================== test session starts ===================================================
platform linux -- Python 3.7.5, pytest-5.4.1, py-1.8.1, pluggy-0.13.1
rootdir: /home/selu/Git/wazuh/api/test/integration, inifile: pytest.ini
plugins: tavern-1.0.0, cov-2.8.1
collected 19 items                                                                                                        

test_rbac_white_manager_endpoints.tavern.yaml ...................                                                   [100%]

==================================================== warnings summary =====================================================
```
```
pytest test_rbac_black_manager_endpoints.tavern.yaml
=================================================== test session starts ===================================================
platform linux -- Python 3.7.5, pytest-5.4.1, py-1.8.1, pluggy-0.13.1
rootdir: /home/selu/Git/wazuh/api/test/integration, inifile: pytest.ini
plugins: tavern-1.0.0, cov-2.8.1
collected 19 items                                                                                                        

test_rbac_black_manager_endpoints.tavern.yaml ...................                                                   [100%]

==================================================== warnings summary =====================================================
```
### Tests cluster and RBAC
```
pytest test_cluster_endpoints.tavern.yaml
=================================================== test session starts ===================================================
platform linux -- Python 3.7.5, pytest-5.4.1, py-1.8.1, pluggy-0.13.1
rootdir: /home/selu/Git/wazuh/api/test/integration, inifile: pytest.ini
plugins: tavern-1.0.0, cov-2.8.1
collected 56 items                                                                                                        

test_cluster_endpoints.tavern.yaml ........................................................                         [100%]

==================================================== warnings summary =====================================================
```
```
pytest test_rbac_white_cluster_endpoints.tavern.yaml
=================================================== test session starts ===================================================
platform linux -- Python 3.7.5, pytest-5.4.1, py-1.8.1, pluggy-0.13.1
rootdir: /home/selu/Git/wazuh/api/test/integration, inifile: pytest.ini
plugins: tavern-1.0.0, cov-2.8.1
collected 21 items                                                                                                        

test_rbac_white_cluster_endpoints.tavern.yaml .....................                                                 [100%]

==================================================== warnings summary =====================================================
```
```
pytest test_rbac_black_cluster_endpoints.tavern.yaml
=================================================== test session starts ===================================================
platform linux -- Python 3.7.5, pytest-5.4.1, py-1.8.1, pluggy-0.13.1
rootdir: /home/selu/Git/wazuh/api/test/integration, inifile: pytest.ini
plugins: tavern-1.0.0, cov-2.8.1
collected 21 items                                                                                                        

test_rbac_black_cluster_endpoints.tavern.yaml .....................                                                 [100%]

==================================================== warnings summary =====================================================
```

## RBAC default roles
```
python3 -m pytest wazuh/rbac/tests/test_default_configuration.py 
============================================ test session starts ============================================
platform linux -- Python 3.7.5, pytest-5.4.1, py-1.8.1, pluggy-0.13.1
rootdir: /home/selu/Git/wazuh/framework
plugins: tavern-1.0.0, cov-2.8.1
collected 28 items                                                                                          

wazuh/rbac/tests/test_default_configuration.py ............................                           [100%]

============================================= warnings summary ==============================================
```

### manager.py
```
python3 -m pytest wazuh/tests/test_manager.py 
=================================================== test session starts ===================================================
platform linux -- Python 3.7.5, pytest-5.4.1, py-1.8.1, pluggy-0.13.1
rootdir: /home/selu/Git/wazuh/framework
plugins: tavern-1.0.0, cov-2.8.1
collected 40 items                                                                                                        

wazuh/tests/test_manager.py ........................................                                                [100%]

==================================================== warnings summary =====================================================
```

### core/manager.py
```
python3 -m pytest wazuh/core/tests/test_manager.py 
=================================================== test session starts ===================================================
platform linux -- Python 3.7.5, pytest-5.4.1, py-1.8.1, pluggy-0.13.1
rootdir: /home/selu/Git/wazuh/framework
plugins: tavern-1.0.0, cov-2.8.1
collected 26 items                                                                                                        

wazuh/core/tests/test_manager.py ..........................                                                         [100%]

==================================================== warnings summary =====================================================
```

## TO DO

- [x] Update RBAC catalogue (waiting for another PR to be merged).
- [x] Write integration tests (basic, RBAC while-list and RBAC black-list)
- [x] Write/update unit tests (manager and core/manager modules)

Kind regards,
Selu.

### CODE REVIEW

File: api/api/controllers/manager_controller.py [hunk 1]
```diff
@@ -357,6 +360,79 @@ async def delete_files(request, pretty=False, wait_for_complete=False, path=None
...
+    :param wait_for_complete: Disable timeout response
+    """
+    try:
+        f_kwargs = {"updated_config": await request.json()}
+    except JSONDecodeError as e:
+        raise_if_exc(APIError(code=2005, details=e.msg))
+
+    dapi = DistributedAPI(f=manager.update_api_config,
+                          f_kwargs=remove_nones_to_dict(f_kwargs),
+                          request_type='local_master',
```
Comment (davidjiglesias) [inline api/api/controllers/manager_controller.py:399]: Shouldn't it be distributed_master and broadcasted?

File: api/api/controllers/manager_controller.py [hunk 2]
```diff
@@ -357,6 +360,79 @@ async def delete_files(request, pretty=False, wait_for_complete=False, path=None
...
+    :param wait_for_complete: Disable timeout response
+    """
+    allowed_fields = {'behind_proxy_server', 'rbac', 'logs', 'cache', 'cors', 'use_only_authd', 'experimental_features'}
+    default_config = {key: configuration.default_configuration[key] for key in allowed_fields}
+
+    f_kwargs = {"updated_config": default_config}
+
+    dapi = DistributedAPI(f=manager.update_api_config,
+                          f_kwargs=remove_nones_to_dict(f_kwargs),
+                          request_type='local_any',
```
Comment (davidjiglesias) [inline api/api/controllers/manager_controller.py:424]: Shouldn't it be distributed_master and broadcasted?

File: api/test/integration/test_cluster_endpoints.tavern.yaml [hunk 3]
```diff
@@ -2532,6 +2532,210 @@ stages:
...
+        Authorization: "Bearer {test_login_token}"
+      method: PUT
+      json:
+        logs:
+          path: "test"
+    response:
+      status_code: 400
+
+  # GET /cluster/api/config
+  - name: Get API configuration that were set above
```
Comment (davidjiglesias) [inline api/test/integration/test_cluster_endpoints.tavern.yaml:2652]:
```suggestion
  - name: Get API configuration that was set above
```

File: framework/wazuh/core/manager.py [hunk 4]
```diff
@@ -213,3 +215,31 @@ def replace_in_comments(original_content, to_be_replaced, replacement):
...
+    """Update dict and subdicts without overriding unspecified keys and write it in the API.yaml file.
+
+    Parameters
+    ----------
+    new_config : dict
+        Dictionary with the new configuration.
+    node_type : str
+        Type of node (master, worker)
+    """
+    if new_config:
```
Comment (davidjiglesias) [inline framework/wazuh/core/manager.py:230]: I don't completely understand the reasoning behind this if else. When can we get an empty new_config?

File: api/api/configuration.py [hunk 5]
```diff
@@ -8,6 +8,8 @@
 from typing import Dict, List, Tuple
 
 import yaml
+
+import copy
```
Comment (AdriiiPRodri) [inline api/api/configuration.py:12]: These imports can be optimized

File: api/api/configuration.py [hunk 6]
```diff
@@ -20,6 +22,42 @@
 from wazuh import common
 
 
+default_configuration = {
+        "host": "0.0.0.0",
```
Comment (AdriiiPRodri) [inline api/api/configuration.py:26]: We can remove a tab (26-58)

File: api/test/integration/env/configurations/rbac/cluster/white_configuration_rbac.sh [hunk 7]
```diff
@@ -2,6 +2,7 @@
...
+#permissions='[{"actions":["cluster:read_config"],"resources":["node:id:master-node","node:id:worker2"],"effect":"allow"},{"actions":["cluster:restart"],"resources":["node:id:worker1"],"effect":"allow"},{"actions":["cluster:status"],"resources":["*"],"effect":"allow"},{"actions":["cluster:read_file"],"resources":["node:id:master-node\&file:path:etc/ossec.conf","node:id:worker2\&file:path:etc/ossec.conf","node:id:master-node\&file:path:etc/rules/local_rules.xml","node:id:worker2\&file:path:etc/rules/local_rules.xml","node:id:master-node\&file:path:ruleset/rules/0350-amazon_rules.xml","node:id:worker2\&file:path:ruleset/rules/0350-amazon_rules.xml"],"effect":"allow"},{"actions":["cluster:upload_file"],"resources":["node:id:master-node","node:id:worker2"],"effect":"allow"},{"actions":["cluster:delete_file"],"resources":["node:id:master-node\&file:path:ruleset/decoders/0005-wazuh_decoders.xml","node:id:worker2\&file:path:etc/ossec.conf"],"effect":"allow"}]'
```
Comment (AdriiiPRodri) [inline api/test/integration/env/configurations/rbac/cluster/white_configuration_rbac.sh:5]: Can we remove this comment?

**Pull request merged**